ui/app/routes/vault/cluster/clients/counts.ts

```diff
@@ -39,38 +37,64 @@ export default class ClientsCountsRoute extends Route {
+  paramOrResponseTimestamp(
+    qpMillisString: string | number | undefined,
+    activityTimeStamp: string | undefined
+  ) {
+    let timestamp: string | undefined;
+    const millis = Number(qpMillisString);
+    if (!isNaN(millis)) {
+      timestamp = fromUnixTime(millis).toISOString();
+    }
+    // fallback to activity timestamp only if there was no query param
+    if (!timestamp && activityTimeStamp) {
+      timestamp = activityTimeStamp;
+    }
+    return timestamp;
+  }
+
+  async getActivity(params: ClientsCountsRouteParams) {
     let activity, activityError;
-    // if there is no start_time we want the user to manually choose a date
-    // in that case bypass the query so that the user isn't stuck viewing the activity error
-    if (start_time) {
+    const query = {
+      // start and end params are optional -- if not provided, will fallback to API default
+      start_time: this.formatTimeQuery(params?.start_time),
+      end_time: this.formatTimeQuery(params?.end_time),
+    };
+    // if CE without start time we want to skip the activity call
+    // so that the user is forced to choose a date range
+    if (this.version.isEnterprise || params.start_time) {
       try {
-        activity = await this.store.queryRecord('clients/activity', {
-          start_time: { timestamp: start_time },
-          end_time: { timestamp: end_time },
-        });
+        activity = await this.store.queryRecord('clients/activity', query);
       } catch (error) {
         activityError = error;
       }
     }
-    return { activity, activityError };
+    return {
+      activity,
+      activityError,
+    };
+  }
+
+  // Takes the string URL param and formats it as the adapter expects it,
+  // if it exists and is valid
+  formatTimeQuery(param: string | number | undefined) {
+    let timeParam: { timestamp: number } | undefined;
+    const millis = Number(param);
+    if (!isNaN(millis)) {
+      timeParam = { timestamp: millis };
+    }
+    return timeParam;
   }
 
   async model(params: ClientsCountsRouteParams) {
     const { config, versionHistory } = this.modelFor('vault.cluster.clients') as ModelFrom<ClientsRoute>;
-    // only enterprise versions will have a relevant billing start date, if null users must select initial start time
-    const startTime = setStartTimeQuery(this.version.isEnterprise, config);
-
-    const startTimestamp = Number(params.start_time) || startTime;
-    const endTimestamp = Number(params.end_time) || getUnixTime(timestamp.now());
-    const { activity, activityError } = await this.getActivity(startTimestamp, endTimestamp);
-
+    const { activity, activityError } = await this.getActivity(params);
     return {
       activity,
       activityError,
       config,
-      endTimestamp,
-      startTimestamp,
+      startTimestamp: this.paramOrResponseTimestamp(params?.start_time, activity?.byMonth[0]?.timestamp),
+      endTimestamp: this.paramOrResponseTimestamp(params?.end_time, activity?.endTime),
```

Comment:
So the new functionality is these params correspond to the range of data _requested_ and the `activity` above (line 93) has start and end time parameters that correspond to the `start_time` and `end_time` from the `/activity` response?

I'm just thinking through the logic we have that lets users know "You have requested [start date] we only have data from [start of data]" 

`paramOrResponseTimestamp` is helpfully named, I wonder if a comment above this function would help clarify the difference between selecting the first month timestamp vs the start/end params on the actual activity response 🤔

Reply:
Good idea, I can definitely add a comment with a description about what this does! 

If query params were used, we want to set the start and end timestamps to match those, as you said, because the API start time matches the first month that data was returned. I found that if I just relied on the API timestamp responses, the message about only having data from [start of data] never showed up, because the timestamps matched. So, instead I had to default to the first month's timestamp if no query param was used, to ensure that the discrepancy message displayed when the data wasn't available for all months.